// @mobilize/snowflake-instrumentation-helpers/lib/RowsWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RowsWrapper = void 0;
var RowsWrapper = /** @class */ (function () {
    function RowsWrapper(rows, logger) {
        this.rows = rows;
        this.index = -1;
        this.logger = logger;
    }
    RowsWrapper.prototype.next = function () {
        this.index = this.index + 1;
        if (this.index >= this.rows.length) {
            return false;
        }
        return true;
    };
    RowsWrapper.prototype.getCurrentRow = function () {
        return this.rows[this.index];
    };
    RowsWrapper.prototype.getColumnValue = function (i) {
        var currentRow = this.getCurrentRow();
        if (!currentRow) {
            this.logger.error('no current row');
            throw new Error('no current row');
        }
        var columnNames = Reflect.ownKeys(currentRow);
        var columnName = columnNames[i - 1];
        return this.rows[this.index][columnName];
    };
    return RowsWrapper;
}());
exports.RowsWrapper = RowsWrapper;
