// @mobilize/snowflake-instrumentation-helpers/lib/SnowflakeWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnowflakeWrapper = void 0;
var StatementWrapper_1 = require("./StatementWrapper");
var SnowflakeWrapper = /** @class */ (function () {
    function SnowflakeWrapper(connection, logger) {
        this.statements = [];
        this.connection = connection;
        this.logger = logger;
    }
    SnowflakeWrapper.prototype.execute = function (json) {
        var stmt = this.createStatement(json);
        stmt.execute();
    };
    SnowflakeWrapper.prototype.createStatement = function (json) {
        var stmt_wrapper = new StatementWrapper_1.StatementWrapper(this.connection, this.logger, json);
        return stmt_wrapper;
    };
    SnowflakeWrapper.prototype.close = function () {
        var that = this;
        this.connection.destroy(function (err, conn) {
            if (err) {
                that.logger.log({
                    level: 'error',
                    message: 'unable to disconnect',
                    reason: err.message
                });
            }
        });
    };
    return SnowflakeWrapper;
}());
exports.SnowflakeWrapper = SnowflakeWrapper;
